polkadot/runtime/parachains/src/assigner_coretime/mod.rs

```diff
@@ -0,0 +1,497 @@
+			Some(q) => q,
+		};
+
+		let next_scheduled = queue.first;
+
+		if next_scheduled > now {
+			// Not yet ready.
+			return
+		}
+
+		// Update is needed:
+		let update = CoreSchedules::<T>::take((next_scheduled, core_idx));
+		let new_first = update.as_ref().and_then(|u| u.next_schedule);
+		descriptor.current_work = update.map(|u| u.into());
+
+		descriptor.queue = new_first.map(|new_first| {
+			QueueDescriptor {
+				first: new_first,
+				// `last` stays unaffected, if not empty:
+				last: queue.last,
+			}
+		});
+	}
+
+	/// Append another assignment for a core.
+	///
+	/// Important only appending is allowed. Meaning, all already existing assignments must have a
+	/// begin smaller than the one passed here. This restriction exists, because it makes the
+	/// insertion O(1) and the author could not think of a reason, why this restriction should be
+	/// causing any problems. Inserting arbitrarily causes a `DispatchError::DisallowedInsert`
+	/// error. This restriction could easily be lifted if need be and in fact an implementation is
+	/// available
+	/// [here](https://github.com/paritytech/polkadot-sdk/pull/1694/commits/c0c23b01fd2830910cde92c11960dad12cdff398#diff-0c85a46e448de79a5452395829986ee8747e17a857c27ab624304987d2dde8baR386).
+	/// The problem is that insertion complexity then depends on the size of the existing queue,
+	/// which makes determining weights hard and could lead to issues like overweight blocks (at
+	/// least in theory).
+	pub fn assign_core(
+		core_idx: CoreIndex,
+		begin: BlockNumberFor<T>,
+		assignments: Vec<(CoreAssignment, PartsOf57600)>,
+		end_hint: Option<BlockNumberFor<T>>,
+	) -> Result<(), DispatchError> {
+		// TODO: Add this assert once the calls `request_core_count` and `notify_core_count`
+		// have been established. assert!(core < core_count);
+
+		// There should be at least one assignment and at most 100
+		ensure!(assignments.len() > 0usize, Error::<T>::AssignmentsEmpty);
+		ensure!(
+			assignments.len() <= MAX_ASSIGNMENTS_PER_SCHEDULE as usize,
+			Error::<T>::TooManyAssignments
+		);
+
+		// Checking for sort and unique manually, since we don't have access to iterator tools.
+		// This way of checking uniqueness only works since we also check sortedness.
+		let assignment_targets = assignments.iter().map(|x| &x.0);
+		let mut maybe_prior = None;
+		for target in assignment_targets {
+			if let Some(prior) = maybe_prior {
+				ensure!(target != prior, Error::<T>::AssignmentsNotUnique);
+				ensure!(target > prior, Error::<T>::AssignmentsNotSorted);
+			}
+			maybe_prior = Some(target);
+		}
+
+		// Check that the total parts between all assignments are equal to 57600
+		let parts_sum = assignments
+			.iter()
+			.map(|assignment| assignment.1)
+			.fold(PartsOf57600::from(0u16), |sum, parts| sum.saturating_add(parts));
+		ensure!(parts_sum <= PartsOf57600::from(57600u16), Error::<T>::OverScheduled);
+		ensure!(parts_sum >= PartsOf57600::from(57600u16), Error::<T>::UnderScheduled);
```

Comment:
@bkchr `?` doesn't work here, I'll make your changes but will keep the `ensure!`. But, woah, there's a bug, because we use `saturating_add` so the overscheduled case is never caught. Will fix that.